spring/src/main/java/io/cloudevents/spring/messaging/CloudEventMessageConverter.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2019-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.cloudevents.spring.messaging;
+
+import java.nio.charset.Charset;
+
+import io.cloudevents.CloudEvent;
+import io.cloudevents.CloudEventContext;
+import io.cloudevents.CloudEventData;
+import io.cloudevents.core.builder.CloudEventBuilder;
+
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.converter.MessageConverter;
+import org.springframework.messaging.support.MessageBuilder;
+
+/**
+ * A {@link MessageConverter} that can translate to and from a {@link Message
+ * Message&lt;byte[]>} or {@link Message Message&lt;String>} and a {@link CloudEvent}. The
+ * {@link CloudEventContext} is canonicalized, with key names given a "ce-" prefix in the
+ * {@link MessageHeaders}.
+ * 
+ * @author Dave Syer
+ *
+ */
+public class CloudEventMessageConverter implements MessageConverter {
+
+	@Override
+	public Object fromMessage(Message<?> message, Class<?> targetClass) {
+		byte[] payload = getBinaryData(message);
+		if (CloudEvent.class.isAssignableFrom(targetClass)
+				&& message.getHeaders().containsKey(CloudEventsHeaders.SPEC_VERSION)) {
+			CloudEventBuilder builder = CloudEventHeaderUtils.fromMap(message.getHeaders());
+			if (payload != null) {
+				return builder.withData(payload).build();
+			}
```

Comment:
Why you build the context with the message reader but the payload not?

Reply:
Because the `CloudEventHeaderUtils` is for headers (i.e. `CloudEventContext`) but the core SDK doesn't allow that separation currently. I noticed the new `CloudEventContextWriter`, so I think that can be used instead of `MessageWriter` in the `toMap()` method. The other direction doesn't work.

Reply:
In this case, what happens if the incoming message is a structured cloudevent message?

Reply:
This doesn't support that use case I guess - it's the same discussion we had about the HTTP converters - Spring has deep knowledge of how to do type conversions and I want to keep that (even if it restricts me to binary cloud event messages). I could look at extending the PR. Or we could put it in a separate one. Which do you prefer?

Reply:
I would prefer to have it in this PR

Reply:
OK. That was fun (see latest commit).